src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Validation.cs

```diff
@@ -720,5 +735,22 @@ internal static unsafe partial class Utf8Utility
             scalarCountAdjustment = tempScalarCountAdjustment;
             return pInputBuffer;
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int AdvSimdMoveMask(Vector128<byte> value)
+        {
+            Debug.Assert(AdvSimd.Arm64.IsSupported);
+
+            // extractedBits[i] = (value[i] & 0x80) == 0x80 & (1 << i);
+            Vector128<byte> mostSignficantBitMask = s_mostSignficantBitMask;
+            Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareEqual(AdvSimd.And(value, mostSignficantBitMask), mostSignficantBitMask);
+            Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitMask128);
```

Comment:
Perhaps if you use the following mask, you can do it in 1 `AddPairwise`.
```csharp
Vector128<byte> mask = Vector128.Create((ushort)0x1001).AsByte();
Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareEqual(AdvSimd.And(value, mostSignficantBitMask), mostSignficantBitMask);
Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, mask);
extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
return extractedBits.AsUInt64().ToScalar();
```

Reply:
I made the static variables local and placed them before the loop.
I also used your suggestion here, which made the second static variable useless.
I noticed your code returns ulong. Maybe I should return uint directly, since that's the value I need where the method is being invoked.

Reply:
Yeah, all you need to see that it is zero or not, so casting to `uint` should be fine.

Reply:
I can't see how this would work, a single application of `AddPairwise` means that the rightmost bits would never be transferred across the lower range of the final `extractedBits` vector.

> Yeah, all you need to see that it is zero or not, so casting to uint should be fine.

If you only care about that, then perhaps using MaxAcross/MinAcross is preferable?

Reply:
With single application of `AddPairwise` you reduce the data from 128-bits to 64-bits and then `.AsUInt64().ToScalar()` gets those 64-bits. You don't care of that value but just if it is zero or not.

> If you only care about that, then perhaps using MaxAcross/MinAcross is preferable?

It is expensive. See https://github.com/dotnet/runtime/pull/37624#issuecomment-641451575.

Reply:
@TamarChristinaArm - Will doing `MinAcross` cheaper than doing 1 compare, 2 and, 1 addpairwise? Basically we want to check if none of the lane of `Vector128<byte>` has `value > 127`.

```csharp
// Vector128<byte> mostSignficantBitMask = Vector128.Create((byte)0x80);
// Vector128<byte> mask = Vector128.Create((ushort)0x1001).AsByte();

Vector128<byte> mostSignificantBitIsSet = AdvSimd.CompareEqual(AdvSimd.And(value, s_mostSignficantBitMask), s_mostSignficantBitMask);
Vector128<byte> extractedBits = AdvSimd.And(mostSignificantBitIsSet, s_bitMask128);
extractedBits = AdvSimd.Arm64.AddPairwise(extractedBits, extractedBits);
return extractedBits.AsUInt64().ToScalar();
```

From the latency calculation it looks like `MinAcross` will be cheaper but can you confirm if getting it through pipeline will be cheaper?

Reply:
Or perhaps
```csharp
Vector128<ulong> extractedBits = AdvSimd.And(value, s_mostSignficantBitMask).AsUInt64();
ulong result1 = extractedBits.GetElement(0);
ulong result2 = extractedBits.GetElement(1);
return result1 + result2;
```

Reply:
> With single application of AddPairwise you reduce the data from 128-bits to 64-bits and then .AsUInt64().ToScalar() gets those 64-bits.

Ok, so proposal is basically to return a 64-bit integer that partitions element masks across nibbles? I think that can work, however I would probably need to refactor a bunch of code, since the mask is being passed around in architecture-agnostic code assuming the x86 layout.

Reply:
So there are 2 cases in which you need `MoveMask` like operation:
1. To see if the MSB of any lane is set. That can be achieved using `MinAcross` or the way I suggest. In my suggestion, I just AND all the lanes with `0x80` and interpret the result as 2 `ulong`. Then just OR those both and check if the result is zero. If yes, then none of the lanes had MSB set to 1. (Again, lets wait until Tamar confirms what is optimal)
2. You need to find out what is the first lane that has specific value as done here:
https://github.com/dotnet/runtime/blob/a0987db4857ea6f15944f4e8306e49517afa5768/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs#L1943) 
Here you need to do `AddPairwise` etc to retain the value. But for 1st case, you don't care what the values are but just need to check that they are zero (or not zero).

Reply:
Yep, I'm firmly in use case 2, effectively passing the mask result to `BitOperations.TrailingZeroCount`. With your proposal I'd further need to divide by 4.

Reply:
> @TamarChristinaArm - Will doing `MinAcross` cheaper than doing 1 compare, 2 and, 1 addpairwise? Basically we want to check if none of the lane of `Vector128<byte>` has `value > 127`.

This is essentially the same question as in https://github.com/dotnet/runtime/pull/38707#discussion_r454032981 right? Reduce the number of elements with a `smaxp`, move it to the genreg side and use `and` to check the top bits.

To answer your original question though, while the `MinAcross` may end up being slightly cheaper latency wise, it would block the pipeline for a quite a long time. So it could slow down things in general in real code. You're right in avoiding it if you can!

Reply:
Just to be clear, I still see 3 `AddPairwise` used here instead of 1 `AddPairwise`. Did I miss any conversation where we decided to use 3 `AddPairwise` operations?

Reply:
I brought back the original code I introduced in the first commit. It seems this code is essentially the same as in @eiriktsarpalis ' [change ](https://github.com/dotnet/runtime/pull/38707/files?file-filters%5B%5D=.cs&hide-deleted-files=true) so I should be able to do the same (one `AddPairwise`) and I should also use the same `bitMask`.